<launch>


    <!-- MASTER NODE! -->
    <node name="run_serial_msckf" pkg="ov_msckf" type="run_serial_msckf" output="screen" clear_params="true" required="true">
<!--    <node name="run_serial_msckf" pkg="ov_msckf" type="run_serial_msckf" output="screen" clear_params="true" required="true" launch-prefix="gdb -ex run &#45;&#45;args">-->

        <!-- bag topics -->
        <param name="topic_imu"      type="string" value="/imu0" />
        <param name="topic_camera0"  type="string" value="/cam0/image_raw" />
        <param name="topic_camera1"  type="string" value="/cam1/image_raw" />

        <!-- bag parameters -->
        <param name="path_bag"    type="string" value="/home/patrick/datasets/tum/dataset-room1_512_16.bag" />
<!--        <param name="path_gt"     type="string" value="$(find ov_data)/tum_vi/dataset-room1_512_16.csv" />-->
        <param name="bag_start"   type="int"    value="1" />
        <param name="bag_durr"    type="int"    value="-1" />

        <!-- world/filter parameters -->
        <param name="use_fej"                type="bool"   value="true" />
        <param name="use_imuavg"             type="bool"   value="true" />
        <param name="use_rk4int"             type="bool"   value="true" />
        <param name="use_stereo"             type="bool"   value="true" />
        <param name="calib_cam_extrinsics"   type="bool"   value="true" />
        <param name="calib_cam_intrinsics"   type="bool"   value="true" />
        <param name="calib_cam_timeoffset"   type="bool"   value="true" />
        <param name="calib_camimu_dt"        type="double" value="0.0" />
        <param name="max_clones"             type="int"    value="11" />
        <param name="max_slam"               type="int"    value="30" />
        <param name="max_slam_in_update"     type="int"    value="999" /> <!-- 15 seems to work well -->
        <param name="max_msckf_in_update"    type="int"    value="999" />
        <param name="max_cameras"            type="int"    value="2" />
        <param name="dt_slam_delay"          type="double" value="3" />
        <param name="init_window_time"       type="double" value="1.0" />
        <param name="init_imu_thresh"        type="double" value="0.70" /> <!-- room1:0.70, room2:0.75, room3:0.70 -->
        <rosparam param="gravity">[0.0,0.0,9.80766]</rosparam>
        <param name="feat_rep_msckf"         type="string" value="GLOBAL_3D" />
        <param name="feat_rep_slam"          type="string" value="ANCHORED_FULL_INVERSE_DEPTH" />
        <param name="feat_rep_aruco"         type="string" value="ANCHORED_FULL_INVERSE_DEPTH" />

        <!-- zero velocity update parameters -->
        <param name="try_zupt"               type="bool"   value="false" />
        <param name="zupt_chi2_multipler"    type="int"    value="2" />
        <param name="zupt_max_velocity"      type="double" value="0.5" />
        <param name="zupt_noise_multiplier"  type="double" value="50" />

        <!-- timing statistics recording -->
        <param name="record_timing_information"   type="bool"   value="false" />
        <param name="record_timing_filepath"      type="string" value="/tmp/ov_msckf_timing.txt" />

        <!-- tracker/extractor properties -->
        <param name="use_klt"            type="bool"   value="true" />
        <param name="num_pts"            type="int"    value="200" />
        <param name="fast_threshold"     type="int"    value="15" />
        <param name="grid_x"             type="int"    value="5" />
        <param name="grid_y"             type="int"    value="5" />
        <param name="min_px_dist"        type="int"    value="5" />
        <param name="knn_ratio"          type="double" value="0.65" />
        <param name="downsample_cameras" type="bool"   value="false" />

        <!-- aruco tag/mapping properties -->
        <param name="use_aruco"        type="bool"   value="false" />
        <param name="num_aruco"        type="int"    value="1024" />
        <param name="downsize_aruco"   type="bool"   value="true" />

        <!-- sensor noise values / update -->
        <param name="up_msckf_sigma_px"            type="double"   value="1" />
        <param name="up_msckf_chi2_multipler"      type="double"   value="1" />
        <param name="up_slam_sigma_px"             type="double"   value="1" />
        <param name="up_slam_chi2_multipler"       type="double"   value="1" />
        <param name="up_aruco_sigma_px"            type="double"   value="1" />
        <param name="up_aruco_chi2_multipler"      type="double"   value="1" />
        <param name="gyroscope_noise_density"      type="double"   value="0.00016" />
        <param name="gyroscope_random_walk"        type="double"   value="0.000022" />
        <param name="accelerometer_noise_density"  type="double"   value="0.0028" />
        <param name="accelerometer_random_walk"    type="double"   value="0.00086" />


        <!-- camera intrinsics -->
        <rosparam param="cam0_wh">[512, 512]</rosparam>
        <rosparam param="cam1_wh">[512, 512]</rosparam>
        <param name="cam0_is_fisheye" type="bool" value="true" />
        <param name="cam1_is_fisheye" type="bool" value="true" />
        <rosparam param="cam0_k">[190.97847715128717, 190.9733070521226, 254.93170605935475, 256.8974428996504]</rosparam>
        <rosparam param="cam0_d">[0.0034823894022493434, 0.0007150348452162257, -0.0020532361418706202, 0.00020293673591811182]</rosparam>
        <rosparam param="cam1_k">[190.44236969414825, 190.4344384721956, 252.59949716835982, 254.91723064636983]</rosparam>
        <rosparam param="cam1_d">[0.0034003170790442797, 0.001766278153469831, -0.00266312569781606, 0.0003299517423931039]</rosparam>


        <!-- camera extrinsics -->
        <rosparam param="T_C0toI">
            [
            -0.99952504,  0.00750192, -0.02989013,  0.04557484,
             0.02961534, -0.03439736, -0.99896935, -0.07116180,
            -0.00852233, -0.99938008,  0.03415885, -0.04468125,
            0.0, 0.0, 0.0, 1.0
            ]
        </rosparam>
        <rosparam param="T_C1toI">
            [
            -0.99951105,  0.00810408, -0.03019914, -0.05545634,
             0.03029912,  0.01251164, -0.99946257, -0.06925002,
            -0.00772188, -0.99988889, -0.01275107, -0.04745286,
            0.0, 0.0, 0.0, 1.0
            ]
        </rosparam>


    </node>


</launch>